File: internal/database/model/template.go
package model

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"time"
	_ "time/tzdata"

	"github.com/paularlott/gossip/hlc"
	"github.com/paularlott/knot/internal/config"

	"github.com/google/uuid"
	"github.com/rs/zerolog/log"
)

const (
	PlatformManual = "manual"
	PlatformDocker = "docker"
	PlatformPodman = "podman"
	PlatformNomad  = "nomad"

	LeafNodeZone = "<leaf-node>"
)

// Template object
type Template struct {
	Id               string                 `json:"template_id" db:"template_id,pk"`
	Name             string                 `json:"name" db:"name"`
	Description      string                 `json:"description" db:"description"`
	Hash             string                 `json:"hash" db:"hash"`
	Platform         string                 `json:"platform" db:"platform"`
	IconURL          string                 `json:"icon_url" db:"icon_url"`
	Job              string                 `json:"job" db:"job"`
	Volumes          string                 `json:"volumes" db:"volumes"`
	Groups           []string               `json:"groups" db:"groups,json"`
	Active           bool                   `json:"active" db:"active"`
	WithTerminal     bool                   `json:"with_terminal" db:"with_terminal"`
	WithVSCodeTunnel bool                   `json:"with_vscode_tunnel" db:"with_vscode_tunnel"`
	WithCodeServer   bool                   `json:"with_code_server" db:"with_code_server"`
	WithSSH          bool                   `json:"with_ssh" db:"with_ssh"`
	WithRunCommand   bool                   `json:"with_run_command" db:"with_run_command"`
	ComputeUnits     uint32                 `json:"compute_units" db:"compute_units"`
	StorageUnits     uint32                 `json:"storage_units" db:"storage_units"`
	ScheduleEnabled  bool                   `json:"schedule_enabled" db:"schedule_enabled"`
	AutoStart        bool                   `json:"auto_start" db:"auto_start"`
	IsDeleted        bool                   `json:"is_deleted" db:"is_deleted"`
	IsManaged        bool                   `json:"is_managed" db:"is_managed"`
	Schedule         []TemplateScheduleDays `json:"schedule" db:"schedule,json"`
	Zones            []string               `json:"zones" db:"zones,json"`
	CustomFields     []TemplateCustomField  `json:"custom_fields" db:"custom_fields,json"`
	MaxUptime        uint32                 `json:"max_uptime" db:"max_uptime"`
	MaxUptimeUnit    string                 `json:"max_uptime_unit" db:"max_uptime_unit"`
	CreatedUserId    string                 `json:"created_user_id" db:"created_user_id"`
	CreatedAt        time.Time              `json:"created_at" db:"created_at"`
	UpdatedUserId    string                 `json:"updated_user_id" db:"updated_user_id"`
	UpdatedAt        hlc.Timestamp          `json:"updated_at" db:"updated_at"`
}

type TemplateScheduleDays struct {
	Enabled bool   `json:"enabled"`
	From    string `json:"from"`
	To      string `json:"to"`
}

type TemplateCustomField struct {
	Name        string `json:"name"`
	Description string `json:"description"`
}

func NewTemplate(
	name string,
	description string,
	job string,
	volumes string,
	userId string,
	groups []string,
	platform string,
	withTerminal bool,
	withVSCodeTunnel bool,
	withCodeServer bool,
	withSSH bool,
	withRunCommand bool,
	computeUnits uint32,
	storageUnits uint32,
	scheduleEnabled bool,
	schedule *[]TemplateScheduleDays,
	zones []string,
	autoStart bool,
	active bool,
	maxUptime uint32,
	maxUptimeUnit string,
	iconURL string,
	customFields []TemplateCustomField,
) *Template {
	id, err := uuid.NewV7()
	if err != nil {
		log.Fatal().Msg(err.Error())
	}

	template := &Template{
		Id:               id.String(),
		Name:             name,
		Description:      description,
		Job:              job,
		Volumes:          volumes,
		Groups:           groups,
		Zones:            zones,
		CreatedUserId:    userId,
		Platform:         platform,
		WithTerminal:     withTerminal,
		WithVSCodeTunnel: withVSCodeTunnel,
		WithCodeServer:   withCodeServer,
		WithSSH:          withSSH,
		WithRunCommand:   withRunCommand,
		ComputeUnits:     computeUnits,
		StorageUnits:     storageUnits,
		CreatedAt:        time.Now().UTC(),
		UpdatedUserId:    userId,
		UpdatedAt:        hlc.Now(),
		IconURL:          iconURL,
		Active:           active,
		MaxUptime:        maxUptime,
		MaxUptimeUnit:    maxUptimeUnit,
		CustomFields:     customFields,
	}
	template.UpdateHash()

	if !scheduleEnabled || schedule == nil {
		template.ScheduleEnabled = false
		template.Schedule = nil
	} else {
		template.ScheduleEnabled = true
		template.Schedule = *schedule
		template.AutoStart = autoStart
	}

	return template
}

func (template *Template) GetVolumes(space *Space, user *User, variables map[string]interface{}) (*CSIVolumes, error) {
	return LoadVolumesFromYaml(template.Volumes, template, space, user, variables)
}

func (template *Template) UpdateHash() {
	hash := md5.Sum([]byte(template.Job + template.Volumes + template.Platform + fmt.Sprintf("%t%t%t%t%t%v", template.WithTerminal, template.WithVSCodeTunnel, template.WithCodeServer, template.WithSSH, template.WithRunCommand, template.CustomFields)))
	template.Hash = hex.EncodeToString(hash[:])
}

func (template *Template) AllowedBySchedule() bool {
	if !template.ScheduleEnabled {
		return true
	}

	// Get the timezone
	cfg := config.GetServerConfig()
	loc, err := time.LoadLocation(cfg.Timezone)
	if err != nil {
		log.Error().Msgf("Error loading timezone: %s", err)
		return false
	}

	// Get the current time and adjust for the server timezone
	now := time.Now().In(loc)

	// Get the day of the week
	dayOfWeek := int(now.Weekday())

	// Get the day schedule, if it doesn't exist then return false
	if len(template.Schedule) <= int(dayOfWeek) {
		return false
	}
	daySchedule := template.Schedule[dayOfWeek]
	if !daySchedule.Enabled {
		return false
	}

	from, err := time.Parse("3:04pm", daySchedule.From)
	if err != nil {
		log.Error().Msgf("Error parsing schedule from time: %s", err)
		return false
	}
	to, err := time.Parse("3:04pm", daySchedule.To)
	if err != nil {
		log.Error().Msgf("Error parsing schedule to time: %s", err)
		return false
	}

	// Test if now is between from and to times
	if now.Hour() >= from.Hour() && now.Hour() <= to.Hour() {
		if now.Hour() == from.Hour() && now.Minute() < from.Minute() {
			return false
		} else if now.Hour() == to.Hour() && now.Minute() > to.Minute() {
			return false
		}

		return true
	}

	return false
}

func (template *Template) IsManual() bool {
	return template.Platform == PlatformManual
}

func (template *Template) IsLocalContainer() bool {
	return template.Platform == PlatformDocker || template.Platform == PlatformPodman
}

// IsValidForZone determines whether the template is valid for deployment in the specified zone.
// The function evaluates zone restrictions based on the template's Zones configuration.
// If no zones are specified, the template is considered valid for all zones.
// Zone names prefixed with '!' are treated as exclusions (negated zones).
// The function first checks for exclusions, then checks for explicit inclusions.
//
// zone is the target zone name to validate against the template's zone restrictions.
//
// Returns true if the template can be deployed in the specified zone, false otherwise.
func (template *Template) IsValidForZone(zone string) bool {
    // If no zones specified, template is valid for all zones
    if len(template.Zones) == 0 {
        return true
    }

    // Check for negated zones first
    for _, z := range template.Zones {
        if z[0] == '!' && z[1:] == zone {
            return false
        }
    }

    // Check for positive zones
    for _, z := range template.Zones {
        if z[0] != '!' && z == zone {
            return true
        }
    }

    return false
}

File: script-apiclient/lib/knot/template.py
# knot.template - Template management library for Knot server

from . import api

def list():
    """List all templates."""
    response = api.get("/api/templates")

    result = []
    for tmpl in response.get("templates", []):
        result.append({
            "id": tmpl.get("template_id"),
            "name": tmpl.get("name"),
            "description": tmpl.get("description", ""),
            "platform": tmpl.get("platform", ""),
            "active": tmpl.get("active", False),
            "usage": tmpl.get("usage", 0),
            "deployed": tmpl.get("deployed", 0)
        })

    return result


def get(template_id):
    """Get template by ID or name."""
    response = api.get(f"/api/templates/{template_id}")
    return _parse_template(response)


def create(name, job="", description="", platform="", volumes="", active=True,
           compute_units=0, storage_units=0, with_terminal=False,
           with_vscode_tunnel=False, with_code_server=False, with_ssh=False,
           with_run_command=False, schedule_enabled=False, icon_url="",
           groups=None, zones=None):
    """Create a new template."""
    body = {
        "name": name,
        "job": job,
        "description": description,
        "platform": platform,
        "volumes": volumes,
        "active": active,
        "compute_units": compute_units,
        "storage_units": storage_units,
        "with_terminal": with_terminal,
        "with_vscode_tunnel": with_vscode_tunnel,
        "with_code_server": with_code_server,
        "with_ssh": with_ssh,
        "with_run_command": with_run_command,
        "schedule_enabled": schedule_enabled,
        "icon_url": icon_url,
        "groups": groups or [],
        "zones": zones or [],
        "schedule": [],
        "custom_fields": []
    }

    response = api.post("/api/templates", body)
    return response.get("template_id")


def update(template_id, name=None, job=None, description=None, platform=None,
           volumes=None, active=None, compute_units=None, storage_units=None,
           with_terminal=None, with_vscode_tunnel=None, with_code_server=None,
           with_ssh=None, with_run_command=None, schedule_enabled=None,
           icon_url=None, groups=None, zones=None):
    """Update template properties."""
    current = api.get(f"/api/templates/{template_id}")

    body = {
        "name": name if name is not None else current.get("name"),
        "job": job if job is not None else current.get("job", ""),
        "description": description if description is not None else current.get("description", ""),
        "platform": platform if platform is not None else current.get("platform", ""),
        "volumes": volumes if volumes is not None else current.get("volumes", ""),
        "active": active if active is not None else current.get("active", True),
        "compute_units": compute_units if compute_units is not None else current.get("compute_units", 0),
        "storage_units": storage_units if storage_units is not None else current.get("storage_units", 0),
        "with_terminal": with_terminal if with_terminal is not None else current.get("with_terminal", False),
        "with_vscode_tunnel": with_vscode_tunnel if with_vscode_tunnel is not None else current.get("with_vscode_tunnel", False),
        "with_code_server": with_code_server if with_code_server is not None else current.get("with_code_server", False),
        "with_ssh": with_ssh if with_ssh is not None else current.get("with_ssh", False),
        "with_run_command": with_run_command if with_run_command is not None else current.get("with_run_command", False),
        "schedule_enabled": schedule_enabled if schedule_enabled is not None else current.get("schedule_enabled", False),
        "icon_url": icon_url if icon_url is not None else current.get("icon_url", ""),
        "groups": groups if groups is not None else current.get("groups", []),
        "zones": zones if zones is not None else current.get("zones", []),
        "schedule": current.get("schedule", []),
        "custom_fields": current.get("custom_fields", []),
        "startup_script_id": current.get("startup_script_id", ""),
        "shutdown_script_id": current.get("shutdown_script_id", ""),
        "auto_start": current.get("auto_start", False),
        "max_uptime": current.get("max_uptime", 0),
        "max_uptime_unit": current.get("max_uptime_unit", "hours")
    }

    api.put(f"/api/templates/{template_id}", body)
    return True


def delete(template_id):
    """Delete a template."""
    api.delete(f"/api/templates/{template_id}")
    return True


def get_icons():
    """Get list of available icons."""
    response = api.get("/api/icons")

    result = []
    for icon in response:
        result.append({
            "description": icon.get("description", ""),
            "source": icon.get("source", ""),
            "url": icon.get("url", "")
        })

    return result


def _parse_template(response):
    """Parse a template response into a standardized dict."""
    schedule = []
    for day in response.get("schedule", []):
        schedule.append({
            "enabled": day.get("enabled", False),
            "from": day.get("from", ""),
            "to": day.get("to", "")
        })

    custom_fields = []
    for cf in response.get("custom_fields", []):
        custom_fields.append({
            "name": cf.get("name", ""),
            "description": cf.get("description", "")
        })

    return {
        "id": response.get("template_id"),
        "name": response.get("name"),
        "description": response.get("description", ""),
        "platform": response.get("platform", ""),
        "job": response.get("job", ""),
        "volumes": response.get("volumes", ""),
        "active": response.get("active", False),
        "is_managed": response.get("is_managed", False),
        "compute_units": response.get("compute_units", 0),
        "storage_units": response.get("storage_units", 0),
        "usage": response.get("usage", 0),
        "deployed": response.get("deployed", 0),
        "hash": response.get("hash", ""),
        "with_terminal": response.get("with_terminal", False),
        "with_vscode_tunnel": response.get("with_vscode_tunnel", False),
        "with_code_server": response.get("with_code_server", False),
        "with_ssh": response.get("with_ssh", False),
        "with_run_command": response.get("with_run_command", False),
        "schedule_enabled": response.get("schedule_enabled", False),
        "auto_start": response.get("auto_start", False),
        "max_uptime": response.get("max_uptime", 0),
        "max_uptime_unit": response.get("max_uptime_unit", "hours"),
        "icon_url": response.get("icon_url", ""),
        "groups": response.get("groups", []),
        "zones": response.get("zones", []),
        "schedule": schedule,
        "custom_fields": custom_fields
    }
